### PR TITLE
[ir] Make sure "StmtFieldManager" to be correct if we modify some fields after the ctor

### PR DESCRIPTION
**Pull request by TH3CHARLie**
Related issue = #1286 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/ir/ir.h [hunk 1]
```diff
@@ -423,14 +423,21 @@ class StmtField {
...
+  }
+
+  explicit StmtFieldNumeric(T value) : value_ptr(nullptr), value(value) {
   }
 
   bool equal(const StmtField *other_generic) const override {
     if (auto other = dynamic_cast<const StmtFieldNumeric *>(other_generic)) {
+      if (other->value_ptr && value_ptr) {
+        return *(other->value_ptr) == *value_ptr;
+      }
```
Comment (xumingkuan) [inline taichi/ir/ir.h:440]: What if only one of them is a pointer?
```suggestion
      } else if (other->value_ptr || value_ptr) {
        return false;
      }
```

File: taichi/ir/ir.h [hunk 2]
```diff
@@ -423,14 +423,21 @@ class StmtField {
 template <typename T>
 class StmtFieldNumeric final : public StmtField {
  private:
+  T *value_ptr;
   T value;
```
Comment (xumingkuan) [inline taichi/ir/ir.h:427]: nit: Consider using `std::variant`?

Comment (TH3CHARLie) [reply taichi/ir/ir.h:427]: Wow, that's something I haven't think of, I'll definitely try that

File: taichi/ir/ir.h [hunk 3]
```diff
@@ -423,15 +423,26 @@ class StmtField {
...
 
   bool equal(const StmtField *other_generic) const override {
     if (auto other = dynamic_cast<const StmtFieldNumeric *>(other_generic)) {
-      return other->value == value;
+      if (std::holds_alternative<T *>(other->value) &&
+          std::holds_alternative<T *>(value)) {
+        return *(std::get<T *>(other->value)) == *(std::get<T *>(value));
+      } else if (std::holds_alternative<T *>(other->value) ||
+                 std::holds_alternative<T *>(value)) {
+        return false;
```
Comment (yuanming-hu) [inline taichi/ir/ir.h:442]:
```suggestion
      } else if (std::holds_alternative<T *>(other->value) ||
                 std::holds_alternative<T *>(value)) {
        TI_ERROR("Inconsistent StmtField value types: a pointer value is compared to a non-pointer value.");
        return false;
```

**Comment by TH3CHARLie** [comment]
> Thank you so much! Looks great.
> 
> If it makes sense, I would let the pointer-nonpointer comparison case return a loud error, since I don't think a normal Taichi program should run into that case. @xumingkuan WDYT?

I also prefer to report error on this. The way taichi construct StmtField would already ensure this case almost never happens.

**Comment by xumingkuan** [comment]
> Thank you so much! Looks great.
> 
> If it makes sense, I would let the pointer-nonpointer comparison case return a loud error, since I don't think a normal Taichi program should run into that case. @xumingkuan WDYT?

Sure, it makes sense to me.

**Pull request merged**